### ISSUE TITLE
Reporter: Second data axis on line graphs

### ISSUE THREAD

**Issue by damies13** [damies13/rfswarm#120]
**Describe the solution you'd like**
Add a second data axis on line graphs, so that you can compare on a single graph things like "users vs average response times" etc

**Comment by damies13** [comment]
This was something also requested by @pkievit86 on the rfswarm Discord

**Comment by damies13** [comment]
Making progress, all the data fields are saving to and being loaded from the report file, graphs don't work yet, they got broken by moving all the original fields to left fields so I could duplicate them for the right fields.
The good news:
- the SQL for left and right field's graph data is auto generating, so hopefully the graphs won't take too long
- This is not a braking change, old v1.0.x reports and report templates should automatically upgrade to left axis graphs, though templates won't auto save in the new format.
The bad news:
- v1.1.x ++ reports and templates won't work on v1.0.x reporter, but this should be understandable

![image](https://user-images.githubusercontent.com/15232871/224480362-8298ca9a-10f0-4736-a3b9-ffb67891322a.png)

**Comment by damies13** [comment]
Today's progress

Dual axis graphs working for the preview, still need to make it work for reports, and still need to make turning axis on and off work.

Still a way to go but, should hopefully be easier from here :crossed_fingers: 

![image](https://user-images.githubusercontent.com/15232871/224542669-5c14c1f9-80e4-4cb9-8e7f-f70fed0861a2.png)

**Comment by pkievit86** [comment]
@damies13 looks great. Is it an idea to built in a setting to change the graphs granularity (data points detail)? And for zooming in we could think of selecting a time range?

**Comment by pkievit86** [comment]
Taken from doc LR Analysis

Enlarging a Section of a Graph
Graphs initially display data representing the entire duration of the scenario. You can enlarge any section of a graph to zoom in on a specific period of the scenario run. For example, if a scenario ran for ten minutes, you can enlarge and focus on the scenario events that occurred between the second and fifth minutes.

To zoom in on a section of the graph:
1 Click inside a graph.
2 Move the mouse pointer to the beginning of the section you want to
enlarge, but not over a line of the graph.
3 Hold down the left mouse button and draw a box around the section you want to enlarge.
4 Release the left mouse button. The section is enlarged.
5 To restore the original view, choose Clear Display Option from the right-
click menu.


Changing the Granularity of the X-Axis

You can make the graphs easier to read and analyze by changing the granularity (scale) of the x-axis. The maximum granularity is half of the graph’s time range. To ensure readability and clarity, the Analysis automatically adjusts the minimum granularity of graphs with ranges of 500 seconds or more.
26

To change the granularity of a graph:
1 Click inside a graph.
2 Select View > Set Granularity, or click the Set Granularity button. The
Granularity dialog box opens.
3 Enter a new granularity value in milliseconds, minutes, seconds, or hours.
4 Click OK.

In the following example, the Hits per Second graph is displayed using different granularities. The y-axis represents the number of hits per second within the granularity interval. 

For a granularity of 1, the y-axis shows the number of hits per second for each one second period of the scenario. 

For a
granularity of 5, the y-axis shows the number of hits per second for every five-second period of the scenario.

Granularity=1
Granularity=5
Granularity=10

In the above graphs, the same scenario results are displayed in a granularity of 1, 5, and 10. The lower the granularity, the more detailed the results. 

For example, using a low granularity as in the upper graph, you see the intervals in which no hits occurred. It is useful to use a higher granularity to study the overall Vuser behavior throughout the scenario.

By viewing the same graph with a higher granularity, you can easily see that overall, there was an average of approximately 1 hit per second

**Comment by damies13** [comment]
> @damies13 looks great. Is it an idea to built in a setting to change the graphs granularity (data points detail)? And for zooming in we could think of selecting a time range?

Hi Patrick, great ideas, but I think the will need to be separate feature requests, I would like to get v1.1.0 out around the end of March, but I'll be struggling to get the features already assigned to the release done by then, and also these ideas would add way too much scope creep to this feature request. It would be helpful if you raise the feature requests and document what you'd like to see in each, as it's useful to have other viewpoints and people to discuss the features with, from your comments I think we need 3 feature requests:

- A setting to change the graphs granularity: Yeah I know about the granularity feature in Loadrunner Analysis, I know how it works, actually one of the things that always frustrated me about LR Analysis is that you could never get graphs of the real data only averaged data, you could set the granularity to 1 but you'd still only get an average of all the points measured in that second. At the moment rfswram-reporter doesn't do any granularity (averaging), it simply displays every result recorded, So this adding this is not simply a matter of adding a field to control the granularity level, it required adding in the code to determine which results fit within the same time window and average them as well, so I guess the FR will be something like "Add an adjustable granularity setting to graphs"
- zooming in we could think of selecting a time range: there's actually 2 parts to this:
  - on the [Global Report Settings](https://github.com/damies13/rfswarm/blob/master/Doc/rfswarm_reporter.md#global-report-settings) screen, you'll see the start and end time are displayed, originally I'd intended to have those as input fields so you could adjust the time period for the whole report, I guess I forgot to do that bit, so a FR to add this in would be good, it should be a not too difficult first step and once the code is in place to add the date ranges to the generated SQL queries, the next part should be easier
  - I had not intended to have a date range constraint for individual graphs, but I can see how it would be useful. if we do it for graphs, we'll also need to do the same for data tables, so the table data matches the graph so this would be the 3rd FR


Just a little background:
When I released the report component with v1.0.0, I knew it was a little rough and quite limited in functionality, it was developed for that release, actually it was the missing functionality that prevented me calling any of the previous releases V1, because I felt it was a necessary part of the solution and one of the biggest weaknesses of the other open source performance test tools. 
Adding the graphs into the manager as part of v0.8.0 was actually a stepping stone to building the reporter, and if you look back over the release history you'll see that was issue #3 and was being worked on in the background while several releases went out, and that has been my approach to the whole development, up to v0.5.0 the focus was on the agent and the communication protocol between the agent and the manager, making it solid and flexibility, that worked and there's been very little change there since, likewise Issue #121 is the first feature/bug that might require a database schema change since release v0.6.3, so the building blocks seem to be solid. 
The reporter in V1.0.0 was always intended to be the building block framework to build a really useful reporting tool, it's quite modular but it hasn't had the time to evolve like the manager and agent have. I just wanted to get it out, let people play with it and find out what features other people wanted to help with prioritizing what got added to the reporter rather than not release it and spend years developing features only I thought would be useful. This FR is actually part of that prioritization process succeeding, so a big thank you for that, 
One of the things i spent the most time on for release v1.1.0 is putting in test automation using github actions, when someone checks in code into a v1.1.0 or higher branch the github actions run robot framework test cases for Agent, Manager, and Reporter on Windows, MaoOS and Linux for python versions 3.7, 3.8, 3.9 3.10 and 3.11, the tests i have in place at the moment are really basic, just run the command line help command and check output contains the component's name and run the GUI of the Manager and Reporter and take a screen shot, the Manager also flips between the tabs and takes screenshots of each tab, I know it's not much in the way of a test, but again is a building block, and hopefully it will enable people to help me add regression tests and we can make future versions of rfswarm really well tested on all platforms, so as you can see I'm trying to build really solid foundations for this tool, I would like some day in the future that it becomes the performance test tool of choice for many testers, and considering I've only had 2 years of part time development vs Loadrunnerr's ~30 years of full time development with many developers, I think I've already achieved a lot, and am looking forward to see what can be achieved with this tool in the next few years. 
My intention always was for the releases following v1.0.0 to focus on adding functionality to Reporter and building out the regression suite as this will need to be the most important priority for the next few releases.

Dave.

**Comment by damies13** [comment]
Also in LR Analysis, you right click on the graph and choose display options, then you need to check the advanced options to get to many of the settings, with advanced options checked or unchecked, the display options, screen is quite messy and had to find things, I would really like to avoid that in rfswarm-reporter, that was part of the decision to have the preview/settings tabs as a core part of the reporter GUI.

**Comment by damies13** [comment]
Successfully generating reports with graphs supporting all the modes, 
- left axis only
- right axis only
- dual axis

Sample reports attached (I can't attach html files here)
[Dual_axis_20230306_183806_demo.docx](https://github.com/damies13/rfswarm/files/10956178/Dual_axis_20230306_183806_demo.docx)
[Dual_axis_20230306_183806_demo.xlsx](https://github.com/damies13/rfswarm/files/10956179/Dual_axis_20230306_183806_demo.xlsx)

[Left_axis_20230306_183806_demo.docx](https://github.com/damies13/rfswarm/files/10956187/Left_axis_20230306_183806_demo.docx)
[Left_axis_20230306_183806_demo.xlsx](https://github.com/damies13/rfswarm/files/10956188/Left_axis_20230306_183806_demo.xlsx)

[Right_axis_20230306_183806_demo.docx](https://github.com/damies13/rfswarm/files/10956192/Right_axis_20230306_183806_demo.docx)
[Right_axis_20230306_183806_demo.xlsx](https://github.com/damies13/rfswarm/files/10956193/Right_axis_20230306_183806_demo.xlsx)

**Comment by damies13** [comment]
Merged into Release V1.1.0 branch

**Comment by damies13** [comment]
> > @damies13 looks great. Is it an idea to built in a setting to change the graphs granularity (data points detail)? And for zooming in we could think of selecting a time range?
> 
> Hi Patrick, great ideas, but I think the will need to be separate feature requests, I would like to get v1.1.0 out around the end of March, but I'll be struggling to get the features already assigned to the release done by then, and also these ideas would add way too much scope creep to this feature request. It would be helpful if you raise the feature requests and document what you'd like to see in each, as it's useful to have other viewpoints and people to discuss the features with, from your comments I think we need 3 feature requests:
> 
>     * A setting to change the graphs granularity: Yeah I know about the granularity feature in Loadrunner Analysis, I know how it works, actually one of the things that always frustrated me about LR Analysis is that you could never get graphs of the real data only averaged data, you could set the granularity to 1 but you'd still only get an average of all the points measured in that second. At the moment rfswram-reporter doesn't do any granularity (averaging), it simply displays every result recorded, So this adding this is not simply a matter of adding a field to control the granularity level, it required adding in the code to determine which results fit within the same time window and average them as well, so I guess the FR will be something like "Add an adjustable granularity setting to graphs"
> 
>     * zooming in we could think of selecting a time range: there's actually 2 parts to this:
>       
>       * on the [Global Report Settings](https://github.com/damies13/rfswarm/blob/master/Doc/rfswarm_reporter.md#global-report-settings) screen, you'll see the start and end time are displayed, originally I'd intended to have those as input fields so you could adjust the time period for the whole report, I guess I forgot to do that bit, so a FR to add this in would be good, it should be a not too difficult first step and once the code is in place to add the date ranges to the generated SQL queries, the next part should be easier
>       * I had not intended to have a date range constraint for individual graphs, but I can see how it would be useful. if we do it for graphs, we'll also need to do the same for data tables, so the table data matches the graph so this would be the 3rd FR
> 
> 
> Just a little background: When I released the report component with v1.0.0, I knew it was a little rough and quite limited in functionality, it was developed for that release, actually it was the missing functionality that prevented me calling any of the previous releases V1, because I felt it was a necessary part of the solution and one of the biggest weaknesses of the other open source performance test tools. Adding the graphs into the manager as part of v0.8.0 was actually a stepping stone to building the reporter, and if you look back over the release history you'll see that was issue #3 and was being worked on in the background while several releases went out, and that has been my approach to the whole development, up to v0.5.0 the focus was on the agent and the communication protocol between the agent and the manager, making it solid and flexibility, that worked and there's been very little change there since, likewise Issue #121 is the first feature/bug that might require a database schema change since release v0.6.3, so the building blocks seem to be solid. The reporter in V1.0.0 was always intended to be the building block framework to build a really useful reporting tool, it's quite modular but it hasn't had the time to evolve like the manager and agent have. I just wanted to get it out, let people play with it and find out what features other people wanted to help with prioritizing what got added to the reporter rather than not release it and spend years developing features only I thought would be useful. This FR is actually part of that prioritization process succeeding, so a big thank you for that, One of the things i spent the most time on for release v1.1.0 is putting in test automation using github actions, when someone checks in code into a v1.1.0 or higher branch the github actions run robot framework test cases for Agent, Manager, and Reporter on Windows, MaoOS and Linux for python versions 3.7, 3.8, 3.9 3.10 and 3.11, the tests i have in place at the moment are really basic, just run the command line help command and check output contains the component's name and run the GUI of the Manager and Reporter and take a screen shot, the Manager also flips between the tabs and takes screenshots of each tab, I know it's not much in the way of a test, but again is a building block, and hopefully it will enable people to help me add regression tests and we can make future versions of rfswarm really well tested on all platforms, so as you can see I'm trying to build really solid foundations for this tool, I would like some day in the future that it becomes the performance test tool of choice for many testers, and considering I've only had 2 years of part time development vs Loadrunnerr's ~30 years of full time development with many developers, I think I've already achieved a lot, and am looking forward to see what can be achieved with this tool in the next few years. My intention always was for the releases following v1.0.0 to focus on adding functionality to Reporter and building out the regression suite as this will need to be the most important priority for the next few releases.
> 
> Dave.

OK I created the feature requests, feel free to add to them if you want to - Dave.

**Comment by damies13** [comment]
As this has been merged into the release v1.1.0 branch, I'll close it now.
